File: JavaScript/Queue.js
// Queues are basically stacks except they go FIFO (First in First out)
let queue = [];

queue.push("ALEX");
queue.push("ALEX2");
queue.push("ALEX3");

// Keep in mind removing from the beginning would reindex everything and would be O(N) time complexity
queue.shift();
// console.log(queue);

// Now from scratch with a LL
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

class Queue {
  constructor() {
    this.first = null;
    this.last = null;
    this.size = 0;
  }

  enqueue(val) {
    const newNode = new Node(val);
    if (!this.first) {
      this.first = newNode;
      this.last = newNode;
    } else {
      this.last.next = newNode;
      this.last = newNode;
    }
    return ++this.size;
  }

  dequeue() {
    if (!this.first) return null;
    const temp = this.first;
    if (this.first === this.last) {
      this.first = null;
    }
    this.first = this.first.next;
    this.size -= 1;
    return temp.val;
  }
}

const q = new Queue();

q.enqueue("1st");
q.enqueue("2nd");
q.enqueue("3rd");

console.log(q);
console.log(q.dequeue());
console.log(q);

File: Python/Queue.py
# Queues are stacks except they are FIFO(First in first out)

queue = []

queue.append("A")
queue.append("B")
queue.append("C")
queue.append("D")
queue.pop(0)
# print(queue)

class Node:
    def __init__(self,val):
        self.val = val
        self.next = None

class Queue:
    def __init__(self):
        self.first = None
        self.last = None
        self.size = 0

    def enqueue(self,val):
        newNode = Node(val)
        if not self.first:
            self.first = newNode
            self.last = newNode
        else:
            self.last.next = newNode
            self.last = newNode
        self.size += 1
        return self.size
    
    def dequeue (self):
        if not self.first: return None

        temp = self.first

        if self.first == self.last:
            self.first = None
        self.first = self.first.next
        self.size -= 1
        return temp.val

q = Queue()

q.enqueue("1")
q.enqueue("2")
q.enqueue("3")
q.enqueue("4")
q.dequeue()
while q.first:
    print(q.first.val)
    q.first = q.first.next